atomic_reactor/utils/operator.py

```diff
@@ -25,6 +27,134 @@
+    _digest = r"(?:@sha256:{base16}{{64}})".format(base16=_base16)
+
+    # A tag is either a named tag or a digest
+    _tag = r"(?:{named_tag}|{digest})".format(named_tag=_named_tag, digest=_digest)
+
+    # Registry is a basic name that contains at least one dot
+    # followed by an optional port number
+    _registry = r"(?:{alnum}{name}*\.{name}*{alnum}(?::\d+)?)".format(alnum=_alnum, name=_name)
+
+    # Namespace is a basic name
+    _namespace = _basic_name
+
+    # Repo is a basic name followed by a tag
+    # NOTE: Tag is REQUIRED, otherwise regex picks up too many false positives,
+    # such as URLs, math and possibly many others.
+    _repo = _basic_name + _tag
+
+    # Pullspec is registry/namespace*/repo
+    _pullspec = r"{registry}/(?:{namespace}/)*{repo}".format(registry=_registry,
+                                                             namespace=_namespace,
+                                                             repo=_repo)
+
+    # Regex for sequence of characters that *could* be a pullspec
+    CANDIDATE = re.compile(r"[a-zA-Z0-9/\-._@:]+")
+
+    # Fully match a single pullspec
+    FULL = re.compile(r"^{pullspec}$".format(pullspec=_pullspec))
+
+    # Find pullspecs in text
+    PULLSPEC = re.compile(_pullspec)
+
+
+def default_pullspec_heuristic(text):
+    """
+    Attempts to find all pullspecs in arbitrary structured/unstructured text.
+    Returns a list of (start, end) tuples such that:
+
+        text[start:end] == <n-th pullspec in text> for all (start, end)
+
+    The basic idea:
+
+    - Find continuous sequences of characters that might appear in a pullspec
+      - That being <alphanumeric> + "/-._@:"
+    - For each such sequence:
+      - Strip non-alphanumeric characters from both ends
+      - Match remainder against the pullspec regex
+
+    Put simply, this heuristic should find anything in the form:
+
+        registry/namespace*/repo:tag
+        registry/namespace*/repo@sha256:digest
+
+    Where registry must contain at least one '.' and all parts follow various
+    restrictions on the format (most typical pullspecs should be caught). Any
+    number of namespaces, including 0, is valid.
+
+    NOTE: Pullspecs without a tag (implicitly :latest) will not be caught.
+    This would produce way too many false positives (and 1 false positive
+    is already too many).
+
+    :param text: Arbitrary blob of text in which to find pullspecs
+    :return: List of (start, end) tuples of substring indices
+    """
+    pullspecs = []
+    for i, j in _pullspec_candidates(text):
+        i, j = _adjust_for_arbitrary_text(text, i, j)
+        candidate = text[i:j]
+        if PullspecRegex.FULL.match(candidate):
+            pullspecs.append((i, j))
+            log.debug("Pullspec heuristic: %s looks like a pullspec", candidate)
+    return pullspecs
+
+
+def _pullspec_candidates(text):
+    return [match.span() for match in PullspecRegex.CANDIDATE.finditer(text)]
```

Comment:
why don't we use PullspecRegex.PULLSPEC here?

which isn't used at all.

so instead of using CANDIDATE just use PULLSPEC

Reply:
also when we do this, we won't need _adjust_for_arbitrary_tex

Reply:
I had the same question, but there are a few use cases that break if the candidate is not used. I think the main reason is so things like `foo.com/repo:latest/foobar` produce no matches instead of `foo.com/repo:latest`.

Reply:
ahh ok ic

Reply:
That's exactly it, I'll add a comment to the PULLSPEC regex (or I could remove that regex, but that would probably leave the same questions)